### ISSUE TITLE
System.ArgumentException: Stream was not readable 

### ISSUE THREAD

**Issue by xantari** [dotnet/runtime#43788]
### Describe the bug

1. Add a configuration via AddJsonStream()
2. Call .Build() on the configuration object
3. Add another configuration via .AddJsonStream()
4. Call .Build() again

You will get a "stream was not readable" issue. It seems the first .Build() will destroy the streams for any further usage, which causes subsequent .Build() calls to fail with "stream was not readable" error.

This is similar to the repro you already received here: dotnet/extensions#1979
Basically what happens is after .Build() is called all the streams are destroyed and the configuration bombs out.

This basically prevents you from putting configuration .json data as embedded resources.

### To Reproduce

Repro is here:
dotnet/extensions#1979

### Expected behavior

Streams should still be able to be read by the IConfigurationBuilder even after multiple .Build() calls.

### Additional context
.NET Core 3.1.7

**Comment by xantari** [comment]
@BrennanConroy Sorry, I haven't had a chance to test NewtonsoftJson yet. Did you still want me to do that? It looks like you already identified the problem...

**Comment by BrennanConroy** [comment]
If it's easy enough to try out, sure.

**Comment by maryamariyan** [comment]
@xantari did you get a chance to test it out?

**Comment by xantari** [comment]
@maryamariyan Nope, looks like BrennanConroy already found the issue so didn't see the point and I had already got rid of that code branch so i'd have to re-do everything.

**Comment by stephentoub** [comment]
@danmosemsft, another example.

**Comment by DL444** [comment]
I would say it might be necessary for Build() method to close the stream so some patterns can work well. For example, in cases where I leave a hook for external code to add additional configurations, closing streams on calling Build() means external parties do not have to worry about closing them themselves. Otherwise, they would have to maintain the streams they opened, and I would have to add another hook after building to notify them so they can close the streams.

**Comment by kdaveid** [comment]
I tested this out with both 
`Microsoft.Extensions.Configuration.Json` (Version 5.0.0.0) as well as 
`Microsoft.Extensions.Configuration.NewtonsoftJson` (Version 5.0.1.0) 

Here is what I did:

```
 WebHost.CreateDefaultBuilder(args)
    .ConfigureAppConfiguration((ctx, builder) =>
    {
        using var securityMem = new System.IO.MemoryStream(System.Text.Encoding.UTF8.GetBytes(JsonSerializer.Serialize(settingsObject)));
        builder.AddNewtonsoftJsonStream(securityMem);
    })
    .UseKestrel();
```

Unfortunately the same results.

If its under my control it works as expected: 
```
using var mem = new MemoryStream(System.Text.Encoding.UTF8.GetBytes(JsonSerializer.Serialize(settingsObject)));
var config = new ConfigurationBuilder().AddJsonStream(mem).Build();
var children = config.GetSection("SettingsKey").GetChildren();    // yay!
```

**Comment by maryamariyan** [comment]
> If we are not supposed to call build more than once, should the framework indicate that as such and throw an error instead?

In general using Build() twice on configuration builder is not in the realm of recommended workflows and how we expect callers to use it. We could add a note in the documentation stating this limitation.

cc: @HaoK @safern

**Comment by broll** [comment]
I solved this by wrapping the stream that seeks to origin on dispose

**Comment by tarekgh** [comment]
@SpringHgui as mentioned in comment https://github.com/dotnet/runtime/issues/43788#issuecomment-807923647 this is not the recommended workflows and how we expect callers to use it. Also, in this thread there is a workaround mentioned in the comment https://github.com/dotnet/runtime/issues/43788#issuecomment-1206670616 too. This makes this issue a low priority compared to the other work we are currently doing.

**Comment by andrii-riabchun** [comment]
I didn't succeed with the approach mentioned in comment [#43788 (comment)](https://github.com/dotnet/runtime/issues/43788#issuecomment-1206670616)
I've managed to solve this problem by implementing a custom `IFileProvider` that stores JSON contents as a byte array in memory:

```cs
internal record InMemoryFileProvider(byte[] Contents) : IFileProvider
{
    public static readonly string FileName = "FILE";

    public IDirectoryContents GetDirectoryContents(string subpath) => new NotFoundDirectoryContents();

    public IFileInfo GetFileInfo(string subpath)
    {
        return subpath == FileName 
            ? new InMemoryFileInfo(FileName, Contents) 
            : new NotFoundFileInfo(subpath);
    }

    public IChangeToken Watch(string filter) => default;
}

internal record InMemoryFileInfo(string FileName, byte[] Data) : IFileInfo
{
    public bool Exists => true;

    public long Length => Data.LongLength;

    public string PhysicalPath => null;

    public string Name => FileName;

    public DateTimeOffset LastModified => DateTimeOffset.MinValue;

    public bool IsDirectory => false;

    public Stream CreateReadStream() => new MemoryStream(Data);
}
```

And usage:
```cs
public static class ConfigurationBuilderExtensions
{
    private static readonly HttpClient _httpClient = new();

    public static IConfigurationBuilder AddDefaultConfiguration(this IConfigurationBuilder config)
    {
        config.AddEnvironmentVariables();
        config.AddUserSecrets<Program>();
        config.AddRemoteJson(config.Build().GetConnectionString("AppConfiguration"));
        return config;
    }

    public static IConfigurationBuilder AddRemoteJson(this IConfigurationBuilder config, string url)
    {
        var response = _httpClient.Send(new(HttpMethod.Get, url));

        var ms = new MemoryStream();
        response.Content.ReadAsStream().CopyTo(ms);

        return config.AddJsonFile(
            new InMemoryFileProvider(ms.ToArray()), 
            InMemoryFileProvider.FileName, 
            false, 
            false);
    }
}
```

**Comment by benmccallum** [comment]
For reference, @broll 's solution would look like this I believe:

```csharp
private sealed class NeverClosingMemoryStream : MemoryStream
{
    protected override void Dispose(bool disposing)
    {
        Seek(0, SeekOrigin.Begin);
    }

    public override ValueTask DisposeAsync()
    {
        Seek(0, SeekOrigin.Begin);
        return default;
    }
}
```

Then you can do the following and if you're host startup has multiple Build calls everything should still be OK:
```
var ms = new NeverClosingMemoryStream();
// Write to the ms
var config = new ConfigurationBuilder()
  .AddJsonStream(ms);
```
